x-pack/plugins/cloud_security_posture/server/plugin.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type {
+  PluginInitializerContext,
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  Logger,
+} from '../../../../src/core/server';
+import { createFindingsIndexTemplate } from './index_template/create_index_template';
+import type {
+  CspServerPluginSetup,
+  CspServerPluginStart,
+  CspServerPluginSetupDeps,
+  CspServerPluginStartDeps,
+} from './types';
+import { defineRoutes } from './routes';
+import { initUiSettings } from './ui_settings';
+
+export class CspPlugin
+  implements
+    Plugin<
+      CspServerPluginSetup,
+      CspServerPluginStart,
+      CspServerPluginSetupDeps,
+      CspServerPluginStartDeps
+    >
+{
+  private readonly logger: Logger;
+  constructor(initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+  }
+
+  public setup(
+    core: CoreSetup<CspServerPluginStartDeps, CspServerPluginStart>,
+    plugins: CspServerPluginSetupDeps
+  ): CspServerPluginSetup {
+    this.logger.debug('csp: Setup');
+    const router = core.http.createRouter();
+
+    // Register server side APIs
+    defineRoutes(router, this.logger);
+    initUiSettings(core.uiSettings);
+
+    return {};
+  }
+
+  public start(core: CoreStart, plugins: CspServerPluginStartDeps): CspServerPluginStart {
+    this.logger.debug('csp: Started');
+    createFindingsIndexTemplate(core.elasticsearch.client.asInternalUser, this.logger).catch(
```

Comment:
I believe that this will create the index template even if the user has not enabled the CSP application. This could cause issues if this code is merged and goes into production. As it is, the default kubebeat pattern (`logs-k8s_cis*`) conflicts with Elastic Security. Perhaps we could chat about this offline.

Reply:
Regarding the index template creation - we removed it, it will be done by our integration, this was a temporary workaround to ease the setup for testing.

@lukeelmers You make a good point, and we have considered using a `kibana.json` configuration entry for enabling/disabling our application. Our reasoning for using an advanced setting instead, is the ability to have our application turned on on cloud instances without touching `kibana.json` and restarting Kibana. This can be useful when we want to demo our application for potential beta customers, on their environment our on "our" existing cloud environments. WDYT?

Reply:
@ari-aviran The use case of being able to enable/disable it on demand via advanced settings instead of requiring a server restart makes sense, however the tradeoff is that it really just affects the client-side plugin -- your server plugin is still going to load and perform side effects like registering http routes. And your `plugin.ts` logic & any dependencies will still get included in the initial plugin bundle on page load for all Kibana deployments. Whereas when taking the config approach, it ensures the plugin is completely disabled on both client & server.

So I guess it comes down to whether the server restart is a dealbreaker for y'all, and if it is, whether you are comfortable taking on the maintenance burden of having some of this code running for all users, regardless of whether they are participating in the beta.

Reply:
Chiming in from a Platform Security perspective--

I also noticed the Advanced Setting "feature flag" and it makes me a bit uneasy.

I haven't had a chance to take a thorough look at this PR, but it's 5k+ lines of code and it's not using Kibana feature privileges yet. The HTTP routes it registers are currently accessible to any authenticated user by default, even if the feature flag is not enabled. And, when you _do_ start using Kibana feature privileges, if your plugin isn't ready for GA, those feature privileges will still be visible in the Role Management section of Kibana, which we'd like to avoid for non-GA'ed features.

It would be safer to have the plugin disabled entirely by using a `kibana.yml` setting as Luke mentioned. Then, your team can iterate on this without worrying about hardening your plugin yet and/or accidentally weakening Kibana's security posture.

A couple things to note:

1. You wouldn't have to edit `kibana.json` (your plugin's manifest) to enable your plugin, Luke is suggesting you use a [`kibana.yml` configuration setting](https://www.elastic.co/guide/en/kibana/current/configuration-service.html). If your plugin registers a boolean config setting named `enabled`, Kibana treats that as a special setting, and if that setting is set to `false` then the entire plugin will be disabled. You'd then enable your plugin by using `xpack.csp.enabled: true` (because your plugin's ID in `kibana.json` is "csp").
2. Note that if you use config settings and you want ESS (Cloud) users to be able to change them, you'll have to add them to the Cloud settings allow-list. It's not difficult but it's in a different repo. If you reach out to me via Slack I can show you an example.

Reply:
@lukeelmers @jportner 
Thanks for chiming in, appreciate the feedback. I changed our plugin from using a UI advanced configuration to using a `kibana.yml` configuration setting.